### ISSUE TITLE
dart:ffi asynchronous C execution

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/sdk#35766]
Calling a computationally expensive C function should be called asynchronously to the Dart mutator thread if the C function does not call back into Dart.

**Comment by fzyzcjy** [comment]
Hi is there any updates? Thanks

**Comment by lumarama** [comment]
have this issue been resolved?

**Comment by lumarama** [comment]
another question: should it be possible to start multiple threads in C function called from Dart/Flutter code via ffi - i.e. to utilize multiple CPU cores?

**Comment by simolus3** [comment]
> should it be possible to start multiple threads in C function called from Dart

This is possible today. The different threads just can't call back into Dart.

**Comment by dcharkes** [comment]
> > should it be possible to start multiple threads in C function called from Dart
> 
> This is possible today. The different threads just can't call back into Dart.

They can, but only through native ports, see the documentation at https://github.com/dart-lang/sdk/issues/37022.

**Comment by Dolaned** [comment]
@dcharkes so this means you would want the original thread that setup the callbacks to make the call back to dart?

**Comment by dcharkes** [comment]
> @dcharkes so this means you would want the original thread that setup the callbacks to make the call back to dart?

Yes, a Dart isolate can have only one thread executing Dart code. Doing a callback on a different thread would break invariants in the Dart execution model having two thread executing Dart code in the same isolate. (One Dart execution thread for example means we can trust that we never have concurrent modification, enabling optimizations.)

Using native ports will schedule a task on the event loop which means that the thread executing Dart code will execute that code when code execution is yielded.

See the samples in https://github.com/dart-lang/sdk/tree/master/samples/ffi/async.

**Comment by Dolaned** [comment]
sorry i guess i was talking about the main thread,
Currently i am setting up the callbacks in dart and passing them to C, but then C has a seperate thread that processes data in the background, when it goes to call those callbacks i setup earlier i get a sigabrt
i thought using dart isolate might fix this problem but it might not?

would i be better off setting up the callbacks in c and then using a messaging pipeline with isolate that communicates with dart?

thanks @dcharkes

**Comment by dcharkes** [comment]
@Dolaned, yes, you're better off using the message mechanism to asynchronous computation.

Using a second isolate has downsides: (1) you can only run 1 task in the background per isolate, so you'd have to have more than one isolate if you want to have more than 1 task running in the background at the same time, and (2) isolates take up some resources (memory, and some startup time).

**Comment by Dolaned** [comment]
@dcharkes thanks for your help, i dont think its related to this though.

we can pass a Pointer<Pointer<Struct>> in a call back right? probably a little off topic now

**Comment by dcharkes** [comment]
In a callback yes, in an asynchronous message no. So you'd pass an `int`:

```c++
  Dart_Port send_port;

  Dart_CObject dart_object;
  dart_object.type = Dart_CObject_kInt64;
  dart_object.value.as_int64 = work_addr;

  const bool result = Dart_PostCObject_DL(send_port, &dart_object);
```

Feel free to open a new issue if you have more questions, or post them on StackOverflow.

**Comment by Dolaned** [comment]
ill open a new issue thanks mate

**Comment by lumarama** [comment]
To avoid dealing with callbacks from C code to Dart, is it possible to call C function from Dart (yes, I know it is), then from C code start a new native thread and exit the main thread - thus returning to Dart, while having that native thread still running in parallel. Then from Dart code start polling the native code to get results back when the native thread is finished?

**Comment by dcharkes** [comment]
@lumarama please use asynchronous callbacks for this use case. In order to not have the Dart thread poll one can use `Dart_PostCObject_DL` to send a message from the C thread to Dart ([C source](https://github.com/dart-lang/sdk/blob/master/runtime/bin/ffi_test/ffi_test_functions_vmspecific.cc), [Dart source](https://github.com/dart-lang/sdk/tree/master/samples/ffi/async)).

> **Update 2020-08-24**: it is possible call a C function synchronously and start some work on another thread and return. (When the work on the other thread is done, an asynchronous callback should notify Dart as in the documentation of #37022.)

**Comment by mkustermann** [comment]
> Calling a computationally expensive C function should be called asynchronously to the Dart mutator thread if the C function does not call back into Dart.
> 
> This issue is tracking a more convenient way do the async call, without the need to do the threading in C.

Moving an expensive C function call off the dart thread is easily possible today via:
```dart
final result = await Isolate.run(() {
  callVeryExpensiveCFunction();
});
```

**Comment by jdiaz97** [comment]
@mkustermann I tried something similar but I'm just getting

`Exception has occurred.
ArgumentError (Invalid argument(s): Illegal argument in isolate message: (object is a Pointer)
 <- Context num_variables: 1
 <- Closure: () => void (from dart:core)
 <- Context num_variables: 4
 <- Context num_variables: 1 parent:{ Context num_variables: 4 }
 <- Closure: () => Null (from dart:core)
 <- Instance of '_RemoteRunner<Null>' (from dart:isolate)
)`

**Comment by simolus3** [comment]
You can't send `Pointer` instances over isolates, but you can send their address and reconstruct them via `Pointer.fromAddress` on the other end.

**Comment by sanjaygholap** [comment]
We call CPP dll in Flutter using Flutter ffi. This is our CPP function in dll.

extern "C" __declspec(dllexport) void StartServerSession (const char* Name,const char* js,int& sessionID,int& ERESULT,void(eventhandler)(const char,int))
export "C"__declspec(dllexport) void InitializeServer(void(eventhandler)(const char,int)){ PtrObj->Initialize(eventhandler); }

This Flutter FFI code 

final DynamicLibrary DllPath = DynamicLibrary.open('path to dll');
   
typedef InitializeFunc = Void Function(Pointer<Void> eventhandler);
typedef InitializeFun = void Function(Pointer<Void> eventhandler);

typedef StartServerSessionC = Void Function(Pointer<Utf8> Name, Pointer<Utf8> js, Pointer<Pointer<Int32>> sessionID, Pointer<Pointer<Int32>> ERESULT,Pointer<Void> eventhandler);
typedef StartServerSession = void Function(Pointer<Utf8> Name, Pointer<Utf8> js, Pointer<Pointer<Int32>> sessionID, Pointer<Pointer<Int32>> ERESULT,Pointer<Void> eventhandler);

void eventHandler(Pointer<Utf8> response, int length) {
  print('Response: ${response.cast<Utf8>.toDartString()}');
}


final initialize = DllPath
    .lookupFunction<InitializeFunc , InitializeFun>('Initialize');

final startServerSession = DllPath
    .lookupFunction<StartServerSessionC, StartServerSession>('StartServerSession');


void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: Text('FFI Demo')),
        body: Center(
          child: ElevatedButton(
            onPressed: () async {
              // Example usage of the FFI functions
                await callStartServerIsolate();
            },
            child: Text('Call DLL Function'),
          ),
        ),
      ),
    );
  }
}

void StartServerFunction(SendPort sendPort) 
{
   print("1");          
   Pointer<Utf8> name = ('path to second dll'.toNativeUtf8());
   Pointer<Utf8> js = {json value}.toNativeUtf8();
              
              // Allocate memory for sessionID and ERESULT
              final sessionID = calloc<Int32>();
          final ERESULT = calloc<Int32>();
            
              
              initialize (Pointer.fromFunction<Void Function(Pointer<Utf8>, Int32)>(eventHandler)as Pointer<Void>);

              // Run startOTXSession on a separate isolate
              startServerSession (
                               name,
                               js,
                               sessionID as Pointer<Pointer<Int32>>,
                               ERESULT as Pointer<Pointer<Int32>>, 
                               Pointer.fromFunction<Void Function(Pointer<Utf8>, Int32)>(eventHandler)as Pointer<Void>);
    
         sendPort.send("task Complete");
}

Future<Void> callStartServerIsolate()async{
   final receivePort = ReceivePort();
final sendPort = receivePort.sendPort;
await Isolate.spawn(StartServerFunction,sendPort)
}
** Showing an error that cannot invoke a native callback outside of an isolate.** Now what happened started SeverSession execution, **but internally in cpp dll they started a separate thread with a callback. **

**Comment by dcharkes** [comment]
> Hi can you please help me to resolved this issue "cannot invoke a native callback outside of an isolate" Now what happened in the C++ DLL file. Internally, they started a separate thread with a callback. so how to handle this issue.[https://stackoverflow.com/questions/78404125/cannot-invoke-a-native-callback-outside-of-an-isolate?noredirect=1#comment138225228_78404125](url)

Hi @sanjaygholap! You need to use `NativeCallable.listener` for callbacks from another thread. (And please open a new issue instead of hijacking existing ones.)

**Comment by sanjaygholap** [comment]
@dcharkes sir , sorry I create new issue
